src/Server/TCPHandler.cpp

```diff
@@ -1178,13 +1178,22 @@ void TCPHandler::receiveQuery()
             throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
         /// TODO: change error code?
 
-        /// initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
-        /// i.e. when the INSERT is done with the global context (w/o user).
-        if (!client_info.initial_user.empty())
+        /// Previously, initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
+        /// (i.e. when the INSERT is done with the global context w/o user),
+        /// but this will make client to use previously authenticated user,
+        /// which is bad (it may have different settings, and even readonly=1)
+        if (client_info.initial_user.empty())
         {
-            LOG_DEBUG(log, "User (initial): {}", client_info.initial_user);
+            LOG_DEBUG(log, "User (no user, interserver mode)");
+            session->resetAuthentication();
```

Comment:
Alternative solution: it seems to solve the issue it's enough to just recreate the session with

```
session = std::make_unique<Session>(server.context(), ClientInfo::Interface::TCP_INTERSERVER);
```

before the line `ClientInfo client_info = session->getClientInfo()` in this function. This alternative solution appears to be much shorter and more logical than reseting authentication.

Reply:
But this will loose all other fields, that is filled on Hello, i.e. `ClientInfo.client_*` fields, and others

Reply:
https://github.com/ClickHouse/ClickHouse/blob/acfa949586808bbc76b8e3423856601d80e6f7ad/src/Server/TCPHandler.cpp#L1137 will replace those fields anyway.

Reply:
Then it is good idea, thanks, done!

Reply:
It is still required to cleanup some fields in `ClientInfo` explicitly - #34263